README.md
# twengtwinkie.github.io
My website


changelog.md
# CHANGELOG

## March 2020

* 13th: Started website, basic navigation.

## April 2020

* 19th: Added favicon, styling, navigation js, git repo, updated .gitignore

js/covid.js
var started = false;
var textEl = document.getElementById("questionText");
var startButtonEl = document.getElementById("startButton");
var yesButtonEl = document.getElementById("yesButton");
var noButtonEl = document.getElementById("noButton");
var yes;

var questions = {
    init: "Do you have a fever?",
    yes: "Are you experiencing shortness of breath?",
    no: "Do you have itchy eyes?",
};

var end = false;
var answers = [];

yesButtonEl.onclick = function () {
    next(true, !end);
    end = true;
};

noButtonEl.onclick = function () {
    next(false, !end);
    end = true;
};

function next(answer, notEnd) {
    if (answer && notEnd) {
        textEl.innerHTML = questions.yes+"<br/><br/>";
        answers.push(answer);
    } else if (notEnd) {
        textEl.innerHTML = questions.no+"<br/><br/>";
        answers.push(answer);
    } else if (!notEnd) {
        answers.push(answer);
        if (answers[0]) {
            if (answers[1]) {
                // Covid
                textEl.innerHTML = "<br/>You MAY have the novel coronavirus. Some other symptoms you may have are <ul><li>Cough</li><li>Fatigue</li><li>Weakness</li><li>Exhaustion</li></ul>";
            } else {
                // Flu
                textEl.innerHTML = "<br/>You MAY have the flu. Some other symptoms you may have are <ul><li>Cough</li><li>Fatigue</li><li>Weakness</li><li>Exhaustion</li></ul>";
            }
        } else {
            if (answers[1]) {
                // Allergies
                textEl.innerHTML = "<br/>You MAY have allergies. Some other symptoms you may have are <ul><li>Sneezing</li><li>Running nose</li></ul>";
            } else {
                // Common cold
                textEl.innerHTML = "<br/>You MAY have the common cold. Some other symptoms you may have are <ul><li>Sneezing</li><li>Running nose</li><li>Mild chest discomfort</li></ul>";
            }
        }
        console.log(answers);
        yesButtonEl.style.display = "none";
        noButtonEl.style.display = "none";
        startButtonEl.style.display = "initial";
        startButtonEl.innerText = "Restart";
        answers = [];
        startButtonEl.onclick = start;
    }
}

function start() {
    end = false;
    textEl.style.display = "initial";
    startButtonEl.style.display = "none";

    yesButtonEl.style.display = "initial";
    noButtonEl.style.display = "initial";

    // main code
    textEl.innerHTML = questions.init+"<br/><br/>";
}
/* 
 */
function main() {
    if (started) {
        // run the program
        start();
    } else {
        textEl.style.display = "none";
        requestAnimationFrame(main);
    }
}

main();

js/layout-format.js
function updateNav() {
    var navBarWidth = document.getElementById("nav-bar").clientWidth - 8;
    var navButtonsList = document.getElementById("nav-bar").children;
    for (let i = 0; i < navButtonsList.length; i++) {
        // This determines the maximum width for each item in the nav-bar
        navButtonsList[i].style.width = ((navBarWidth + 4 * (-4 * navButtonsList.length + 1)) / navButtonsList.length + "px");
    }
}

function updateLayout() {
    var footerWidth = document.getElementsByTagName("body")[0].clientWidth - 21;
    var articleWidth = 66 * (footerWidth + 21) / 100;
    var asideWidth = 34 * (footerWidth + 21) / 100 - 4;
    document.getElementsByTagName("article")[0].style.width = articleWidth + "px";
    document.getElementsByTagName("aside")[0].style.width = asideWidth + "px";
    if (document.getElementsByTagName("article")[0].clientWidth < 650) {
        document.getElementsByTagName("article")[0].style.width = footerWidth + 21 + "px";
        document.getElementsByTagName("aside")[0].style.width = footerWidth + 21 + "px";
        document.getElementsByTagName("article")[0].style.float = "none";
        document.getElementsByTagName("aside")[0].style.float = "none";
    } else {
        document.getElementsByTagName("article")[0].style.float = "left";
        document.getElementsByTagName("aside")[0].style.float = "right";
    }
}
function updateLoop() {
    updateNav();
    updateLayout();
    requestAnimationFrame(updateLoop);
}

updateLoop();

space-exploration/main.js
var canvas = /** @type {HTMLCanvasElement} */ document.createElement("canvas");
canvas.width = document.getElementsByTagName("body")[0].clientWidth * 0.6;
canvas.height = 9 * canvas.width / 16;
var coordwidth = 1000;
var coordheight = 562.5; // dynamic canvas scaling.

document.getElementById("canvasBackDiv").appendChild(canvas);
if (canvas.getContext) {
    /** @type {CanvasRenderingContext2D} */
    var c = canvas.getContext('2d');
}
var mainhdl;
var enemies = {};
var bullets = {};
var player = new Player(coordwidth / 6, 5 * coordheight / 6, canvas, bullets, enemies);
var HUDobj = new HUD(c);
var mute = false;

// main
var endless = function () {
    endlessb.style.display = "none";
    campaignb.style.display = "none";
    // resize canvas accordingly
    canvas.width = document.getElementsByTagName("body")[0].clientWidth * 0.6;
    canvas.height = 9 * canvas.width / 16;
    // draw background
    c.clearRect(0, 0, canvas.width, canvas.height);
    var img = document.createElement("img");
    img.src = "images/background.jpg";
    c.drawImage(img, 0, 0, canvas.width, canvas.height);
    // mute audio
    if (mute) {
        lasersound.volume = 0;
        killsound.volume = 0;
        losesound.volume = 0;
    } else {
        lasersound.volume = 0.25;
        killsound.volume = 0.5;
        losesound.volume = 1;
    }
    try {
        /*for (let _ in Object.keys(bullets)) {
            bullets[Object.keys(bullets)[_]].update();
            if (bullets[Object.keys(bullets)[_]].color == "Red") {
                alert("It has been found.");
            }
        }*/
        for (let i = 0; i < Object.keys(bullets).length; i++) {
            bullets[Object.keys(bullets)[i]].update();
        }

        for (let i = 0; i < Object.keys(enemies).length; i++) {
            enemies[Object.keys(enemies)[i]].update();
        }
        player.update();
    } catch (TypeError) {}

    if (Math.floor(Math.random() * 100) == 1) {
        var x = Math.random();
        enemies[x] = (new Enemy(Math.random() * coordwidth, Math.random() * coordheight, canvas, bullets, enemies, player, x));
    }
    // HUD
    c.fillStyle = "#ff8b26";
    c.strokeStyle = "#ff005d";
    c.lineWidth = 4;
    c.strokeRect(10, 10, 500 * canvas.width / coordwidth, 20 * canvas.height / coordheight);
    c.fillRect(10, 10, player.health * 5 * canvas.width / coordwidth, 20 * canvas.height / coordheight);

    HUDobj.display();
    // recursively call the next frame
    mainhdl = requestAnimationFrame(endless);
};


var endlessb = document.createElement("img");
endlessb.src = "./images/UI/endless.png";
endlessb.style.position = "absolute";
endlessb.style.width = 222 * canvas.width / 550 + "px";
endlessb.style.height = 39 * canvas.width / 550 + "px";
endlessb.style.top = ((canvas.getBoundingClientRect().top + canvas.height / 2) - 39 * canvas.width / 1100) + 50 + "px";
endlessb.style.left = ((canvas.getBoundingClientRect().left + canvas.width / 2) - 222 * canvas.width / 1100) + "px";
document.body.appendChild(endlessb);

var campaignb = document.createElement("img");
campaignb.src = "./images/UI/campaign.png";
campaignb.style.position = "absolute";
campaignb.style.width = 222 * canvas.width / 550 + "px";
campaignb.style.height = 39 * canvas.width / 550 + "px";
campaignb.style.top = ((canvas.getBoundingClientRect().top + canvas.height / 2) - 39 * canvas.width / 1100) - 50 + "px";
campaignb.style.left = ((canvas.getBoundingClientRect().left + canvas.width / 2) - 222 * canvas.width / 1100) + "px";
document.body.appendChild(campaignb);

endlessb.addEventListener("click", endless);

projects/projects-navbar-content.js
function set(xhttp, hdl) {
    xhttp.onreadystatechange = hdl;
}

function includeHTML() {
    var z, i, elmnt, file, xhttp;
    var func = function () {
        if (this.readyState == 4) {
            if (this.status == 200) {
                elmnt.innerHTML = this.responseText;
            }
            if (this.status == 404) {
                elmnt.innerHTML = "Page not found.";
            }
            /*remove the attribute, and call this function once more:*/
            elmnt.removeAttribute("include-html");
            includeHTML();
        }
    };
    /*loop through a collection of all HTML elements:*/
    z = document.getElementsByTagName("*");
    for (i = 0; i < z.length; i++) {
        elmnt = z[i];
        /*search for elements with a certain atrribute:*/
        file = elmnt.getAttribute("include-html");
        if (file) {
            /*make an HTTP request using the attribute value as the file name:*/
            xhttp = new XMLHttpRequest();
            set(xhttp, func);
            xhttp.open("GET", file, true);
            xhttp.send();
            /*exit the function:*/
            return;
        }
    }
}

includeHTML();

space-exploration/classes.js
function handleCircleCollision(key1, obj1, key2, obj2) {
    // comparision is ALWWAYS bullet is 1, enemy/player is 2
    var x1, x2, y1, y2, r1, r2;
    x1 = (obj1[key1].x + 14 * Math.cos(obj1[key1].rad)) * obj2[key2].scale;
    y1 = (obj1[key1].y + 14 * Math.sin(obj1[key1].rad)) * obj2[key2].scale;
    r1 = obj1[key1].boxradius * obj2[key2].scale ** 2;
    x2 = obj2[key2].x * obj2[key2].scale;
    y2 = obj2[key2].y * obj2[key2].scale;
    r2 = obj2[key2].boxradius * obj2[key2].scale ** 2;
    if (Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2) <= r1 + r2) {
        // return true;
        // confirmed collision
        if (player.health <= 0) {
            losesound.play();
            console.log("We have a loser!");
            var temp = obj2[key2];
            var finalscore = obj2[key2].score;
            obj2[key2].ctx.fillStyle = "black";
            obj2[key2].ctx.fillRect(0, 0, obj2[key2].canvas.width, obj2[key2].canvas.height);
            obj2[key2].ctx.font = "28px Ken Vector Future";
            obj2[key2].ctx.textAlign = "center";
            obj2[key2].ctx.fillStyle = "white";
            obj2[key2].ctx.fillText("You have lost", obj2[key2].canvas.width / 2, obj2[key2].canvas.height / 2);

            var s = function () {
                temp.ctx.fillStyle = "black";
                temp.ctx.fillRect(0, 0, obj2[key2].canvas.width, obj2[key2].canvas.height);
                temp.ctx.font = "48px Ken Vector Future";
                temp.ctx.textAlign = "center";
                temp.ctx.fillStyle = "red";
                temp.ctx.fillText("You have lost", obj2[key2].canvas.width / 2, obj2[key2].canvas.height / 2 - 18);
                temp.ctx.font = "24px Ken Vector Future";
                temp.ctx.fillStyle = "white";
                temp.ctx.fillText("Your final score was: " + finalscore, obj2[key2].canvas.width / 2, obj2[key2].canvas.height / 2 + 18);
                window.cancelAnimationFrame(mainhdl);
                window.requestAnimationFrame(s);
            };
            s();
        } else {
            player.health -= 5;
            delete obj1[key1];
        }
    }
}

class Ship {
    /**
     * Create an Ship object
     * @param sx Starting x position
     * @param sy Starting y position
     * @param canvas The canvas element
     */
    constructor(sx, sy, canvas) {
        {
            /** @type {CanvasRenderingContext2D} */
            this.ctx = canvas.getContext("2d");

            this.canvas = canvas;
        }
        /* Initializes the ship. sx is the starting x, sy is the starting y. */
        this.x = sx;
        this.y = sy;
        this.deg = 0; // 0 is up, 90 is right, 180 is bottom, 270 is left
        this.image = document.createElement("img");
        this.dx = 0;
        this.dy = 0;
        this.agility = 0.1;
        this.speed = 3;
        this.turningRadius = 12 / this.speed;
        this.scale = canvas.width / (40 * this.width);
        this.reload = true;
    }
    /**
     * Updates the object
     */
    update() {
        this.scale = canvas.width / (40 * this.width);
        this.deg = this.deg % 360;
        this.rad = Math.PI * (this.deg - 90) / 180;
        if (this.image.complete) {
            this.x += this.dx;
            this.y += this.dy;
            // no drag in space!!! [very little]
            this.dx /= 1.001;
            this.dy /= 1.001;
            // rotate the image
            // Save the current context  
            this.ctx.save();
            this.ctx.translate(this.x * this.canvas.width / 1000, this.y * this.canvas.height / coordheight);
            this.ctx.rotate(this.rad + Math.PI / 2);
            this.ctx.drawImage(this.image, -this.width * this.scale / 2, -this.height * this.scale / 2, this.width * this.scale, this.height * this.scale);
            // And restore the context ready for the next loop 
            this.ctx.restore();

            // wrapping screen
            if (this.x >= 1000) {
                this.x = 0;
            }
            if (this.x < 0) {
                this.x = 1000;
            }
            if (this.y >= 562.5) {
                this.y = 0;
            }
            if (this.y < 0) {
                this.y = 562.5;
            }
        }
    }
    /**
     * Shoots a bullet
     * @param {String} color The color of the bullet. Can be "Green", "Red", or "Blue"
     */
    shoot(bullets, enemies, player, color) {
        let x = Math.random(); // any key is fine, but the keys must stay constant, so use object with random keys. 
        // You need 5 * 10 ^ 15 entities for that to overlap
        if (color == "Red") {
            bullets[x] = new Bullet(color, this.x, this.y, (this.deg + 180) % 360, this.canvas, bullets, enemies, player, x);
        } else {
            bullets[x] = new Bullet(color, this.x, this.y, this.deg, this.canvas, bullets, enemies, player, x);
        }
    }

    /**
     * Acceleration
     */
    accelerate() {
        if (Math.sqrt(this.dx * this.dx + this.dy * this.dy) < this.speed) {
            this.dx += this.agility * Math.cos(this.rad);
            this.dy += this.agility * Math.sin(this.rad);
        }
    }
    /**
     * Turn left
     */
    left() {
        this.deg -= this.turningRadius;
    }
    /**
     * Turn right
     */
    right() {
        this.deg += this.turningRadius;
    }
}

class Player extends Ship {
    /**
     * Create a player object
     * @param sx Starting x position
     * @param sy Starting y position
     * @param canvas The canvas element
     * @param bullets The bullets object
     * @param enemies The enemies object
     */
    constructor(sx, sy, canvas, bullets, enemies) {
        super(sx, sy, canvas);
        this.image.src = "images/playerShip2_blue.png";
        this.width = 112;
        this.height = 75;
        this.bullets = bullets;
        this.enemies = enemies;
        this.boxradius = 40;
        this.score = 0;
        this.agility = 0.3;
        this.speed = 5;
        this.health = 100;
    }
    update() {
        super.update();
        this.health += 0.02;
        if (this.health > 100) {
            this.health = 100;
        }
        this.keys();
    }
    keys() {
        if (map.a || map.A || map.ArrowLeft) {
            this.left();
        }
        if (map.d || map.D || map.ArrowRight) {
            this.right();
        }
        if (map.w || map.W || map.ArrowUp) {
            this.accelerate();
        }
        if (map[" "] && this.reload == true) {
            this.reload = false;
            this.shoot(this.bullets, this.enemies, this, "Blue");
            var me = this;
            window.setTimeout(function () {
                me.reload = true;
            }, 750);
        }
    }
}

class Enemy extends Ship {
    /**
     * Create an enemy object
     * @param sx Starting x position
     * @param sy Starting y position
     * @param canvas The canvas element
     */
    constructor(sx, sy, canvas, bullets, enemies, player, mykey) {
        super(sx, sy, canvas);
        switch (Math.floor(Math.random() * 5) + 1) {
            case 1:
                this.image.src = "images/Enemies/enemyBlack1.png";
                this.width = 93;
                this.speed = 4;
                break;
            case 2:
                this.image.src = "images/Enemies/enemyBlack2.png";
                this.width = 104;
                this.speed = 3;
                break;
            case 3:
                this.image.src = "images/Enemies/enemyBlack3.png";
                this.width = 103;
                this.speed = 3;
                break;
            case 4:
                this.image.src = "images/Enemies/enemyBlack4.png";
                this.width = 82;
                this.speed = 1;
                break;
            case 5:
                this.image.src = "images/Enemies/enemyBlack5.png";
                this.width = 97;
                this.speed = 2;
                break;
        }
        this.height = 84;
        this.circleRadius = Math.random() * 50;
        this.circleMovement = Math.random() * 2;
        this.circling = false;
        this.boxradius = 20;
        this.bullets = bullets;
        this.enemies = enemies;
        this.mykey = mykey;
        this.player = player;
    }
    /**
     * Updates the enemy
     */
    update() {
        super.update();
        this.scale = canvas.width / (40 * this.width);
        this.track();
    }
    /**
     * Track the player and follow
     */
    track() {
        this.diffx = this.x - player.x;
        this.diffy = this.y - player.y;
        this.targetAngle = 180 + (180 * Math.atan2(this.diffy, this.diffx) / Math.PI + 270) % 360;

        if (Math.abs(this.deg - this.targetAngle) < 10) {
            this.deg = this.targetAngle;
        } else if (Math.abs(this.deg - this.targetAngle) > 350) {
            this.deg = this.targetAngle;
        }

        if (Math.abs((this.deg - 1) - this.targetAngle) % 360 < Math.abs(this.deg - this.targetAngle) % 360) {
            /**
             * If it gets you closer, do it
             */
            this.left();
        } else {
            this.right();
        }

        let i = 0;
        while ((this.diffx ** 2 + this.diffy ** 2) ** (1 / 2) < 100 + this.circleRadius && i < 4) {
            this.x -= Math.sign(this.dx);
            this.y -= Math.sign(this.dy);
            this.diffx = this.x - player.x;
            this.diffy = this.y - player.y;
            i++; // safety of CPU protection :)
        }

        if ((this.diffx ** 2 + this.diffy ** 2) ** (1 / 2) >= 100 + this.circleRadius) {
            if (Math.sqrt(this.dx * this.dx + this.dy * this.dy) < this.speed) {
                if (!isNaN(0.05 * Math.cos(this.rad))) {
                    this.dx += -0.05 * Math.cos(this.rad);
                }
                if (!isNaN(0.05 * Math.sin(this.rad))) {
                    this.dy += -0.105 * Math.sin(this.rad);
                }
            }
        }
        if (100 + this.circleRadius - 2 < (this.diffx ** 2 + this.diffy ** 2) ** (1 / 2) < 100 + this.circleRadius + 2) {
            // in a circle orbit
            if (!this.circling) {
                this.dx += this.circleMovement * Math.cos(Math.PI * (90 + this.deg) / 180);
                this.dy += this.circleMovement * Math.sin(Math.PI * (90 + this.deg) / 180);
            }
            this.circling = true;
        } else {
            this.circling = false;
        }
        var me;
        if (this.reload == true) {
            this.reload = false;
            this.shoot(this.bullets, this.enemies, this.player, "Red");
            me = this;
            window.setTimeout(function () {
                me.reload = true;
            }, 2000);
        }
    }
}
class Bullet {
    constructor(color, x, y, deg, canvas, bullets, enemies, player, mykey) {
        this.color = color;
        this.player = player;
        this.x = x;
        this.y = y;
        this.mykey = mykey;
        this.deg = deg;
        this.rad = Math.PI * (this.deg - 90) / 180;
        this.canvas = canvas;
        this.ctx = canvas.getContext("2d");
        this.image = document.createElement("img");
        this.color = color;
        this.image.src = "images/Lasers/laser" + color + "07.png";
        this.width = 9;
        this.height = 37;
        this.boxradius = 5; // will be shifted to top
        // center (x,y) with be 14 pixels shifted up.
        // outsider function will check
        // 14*cos(deg) = xcomponent
        // 14*sin(deg) = ycomponent
        this.dx = 0;
        this.dy = 0;
        this.ticks = 0;
        this.bullets = bullets;
        this.enemies = enemies;
        this.scale = canvas.width / (400 * this.width);

        if (this.color == "Blue" || this.color == "Green") {
            lasersound = document.createElement("audio");
            lasersound.src = "./audio/sfx_laser" + 2 + ".ogg";
            lasersound.volume = 0.25;
            lasersound.play();
        }
    }
    update() {
        if (this.image.complete) {
            this.ticks += 1;
            this.scale = canvas.width / (400 * this.width);
            this.dx = 7 * Math.cos(this.rad);
            this.dy = 7 * Math.sin(this.rad);

            this.ctx.save();
            this.ctx.translate(this.x * this.canvas.width / 1000, this.y * this.canvas.height / coordheight);
            this.ctx.rotate(this.rad + Math.PI / 2);
            this.ctx.drawImage(this.image, -this.width * this.scale / 2, -this.height * this.scale / 2, this.width * this.scale, this.height * this.scale);
            // And restore the context ready for the next loop 
            this.ctx.restore();

            this.x += this.dx;
            this.y += this.dy;

            if (this.x >= 1000) {
                this.x = 0;
            }
            if (this.x < 0) {
                this.x = 1000;
            }
            if (this.y >= 562.5) {
                this.y = 0;
            }
            if (this.y < 0) {
                this.y = 562.5;
            }
        }
        if (this.color == "Green" || this.color == "Blue") {
            for (let _ in Object.keys(this.enemies)) { // broad sweeps
            }
            for (let i = 0; i < Object.keys(this.enemies).length; i++) {
                let COVID19 = this.enemies[Object.keys(this.enemies)[i]];
                if (Math.sqrt((COVID19.x - this.x) ** 2 + (COVID19.y - this.y) ** 2) < COVID19.boxradius) { // broad sweeps
                    delete bullets[this.mykey];
                    delete enemies[COVID19.mykey];
                        killsound = document.createElement("audio");
                        killsound.src = "./audio/sfx_twoTone.ogg";
                        killsound.volume = 0.5;
                        killsound.play();
                    
                    this.player.score += 100;
                }
            }
        } else if ((this.player.x - this.x) < 30 && (this.player.y - this.y) < 30) { // broad sweeps
            handleCircleCollision(this.mykey, this.bullets, 0, {
                0: this.player,
            }); // surgical strikes
        }

        if (this.ticks > 100) {
            delete this.bullets[this.mykey];
        }
    }
}

var lasersound = document.createElement("audio");
lasersound.src = "./audio/sfx_laser" + 2 + ".ogg";
lasersound.volume = 0.25;

var killsound = document.createElement("audio");
killsound.src = "./audio/sfx_twoTone.ogg";
killsound.volume = 0.5;

var losesound = document.createElement("audio");
losesound.src = "./audio/sfx_lose.ogg";


class UI {

}

class HUD {
    constructor(context) {
        this.c = context;
        this.numImgList = [];
        for (let i = 0; i < 10; i++) {
            var thing = document.createElement("img");
            thing.src = "./images/UI/numeral" + i + ".png";
            this.numImgList.push(thing);
        }
        console.log(thing);
    }
    display() {
        this.displayScore(player.score);
        this.displayHealth(player.health);
        this.displayMuteButton();
    }
    displayScore(score) {
        var scoreArr = (score + "").split("");
        for (let i = 0; i < scoreArr.length; i++) {
            this.c.drawImage(this.numImgList[scoreArr[i]], canvas.width - 19 * (scoreArr.length - i) - 10, 10);
        }
    }
    displayHealth(health) {
        c.fillStyle = "#ff8b26";
        c.strokeStyle = "#ff005d";
        c.lineWidth = 4;
        c.strokeRect(10, 10, 500 * canvas.width / coordwidth, 20 * canvas.height / coordheight);

        c.fillRect(12, 12, health * 5 * canvas.width / coordwidth - 4, 20 * canvas.height / coordheight - 4);
    }
    displayMuteButton() {

    }
}